### PR TITLE
Adding Charged Fragment Types to MS2 Model Weights  

### PR DESCRIPTION
**Pull request by mo-sameh**
This PR introduces a small but useful modification to the **MS2 model interface**  and **`ModelMS2Bert`**,by including **charged fragment types** as an attribute within the model weights. With this change, charged fragment types are saved and loaded from disk with the weights, making it easier to adjust the requested fragment types for prediction and training.  

## Main Benefits  
- **More flexibility in training and inference**  
  - Supports **partial weight loading** to continue training with different fragment types (as suggested in **PR #226**).  
  - Allows prediction with **any subset of supported fragment types**, including masking modloss fragments—removing the need for a separate `mask_modloss` argument.  
- **Better handling of unsupported fragment types**  
  - Now raises clear and interpretable errors when an invalid fragment type is requested.  

I've added a notebook (**`adapt_charged_fragtypes.ipynb`**) that walks through different use cases, including compatibility with older weight formats. It would be great if you could take a look and let me know if anything is missing.  

This update modifies **`ModelMS2Bert`**, but if this approach looks good, I’d be happy to extend it to other architectures in a follow-up PR. 

Supported use cases with the new format:
| Fragtypes use case              | Override from weights (*) | Safe to predict | Safe to train |
|---------------------------------|--------------------------|----------------|--------------|
| requested = supported   (1)       |         False                 |      ✅          |      ✅        |
| requested ⊆ supported    (2)     |            False              |         ✅      |       ❌       |
| requested ⊈ supported     (3)    |             False             |        ❌        |     ❌         |
|                  Any               |            True              |       ✅         |      ✅        |

(1) The ideal use case (only one supported by the old implementation) where users know and request exactly the same farg types supported in the model weights.
(2) Users only need to predict a subset of the frag types supported by the loaded weights. 
(3) Users requested charged frag types that are not supported (can easily be identified now since we don't only look at the number of requested frag types).
(*) `Override from weights` is the new argument added to the MS2 model, this allow users to load models without knowing exactly what are the supported frag types in a pretrained model. So this overrides the requested frag types and uses all supported frag types by the loaded model.

Safe to train: Any model is automatically set to 'safe to train' when the `train` function is called-- by modifying *only* the underlying model output head to align with the requested frag types. After training, the model will automatically be in a 'safe to predict' state.

### CODE REVIEW

File: nbs_trials/adapt_charged_fragtypes.ipynb [hunk 1]
```diff
@@ -0,0 +1,1446 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "legacy_path = \"../legacy_pretrained_models/generic/ms2.pth\"\n",
+    "new_path = \"../new_pretrained_models/generic/ms2.pth\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
```
Comment (GeorgWa) [inline nbs_trials/adapt_charged_fragtypes.ipynb:202]: I would suggest to make more clear, that we will first show the old format and how the old weight format has lead to issued:

So three examples:
1. Request `charged_frag_types` as in weights and loading goes fine
2. Request `charged_frag_types` with different shape and loading fails with weight missmatch
3. Request `charged_frag_types` with same shape (but different types or order) and the model pretends it works but is acually broke (worst case).

File: nbs_trials/adapt_charged_fragtypes.ipynb [hunk 2]
```diff
@@ -0,0 +1,1446 @@
...
+    "it should the save the new weights in the new path.\n",
+    "\"\"\"\n",
+    "# model.save(new_path)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## User importing weighst in the new format  "
```
Comment (GeorgWa) [inline nbs_trials/adapt_charged_fragtypes.ipynb:541]: Here I would be more clear. Benefits of the new parameter format shown with different use-cases.

File: nbs_trials/adapt_charged_fragtypes.ipynb [hunk 3]
```diff
@@ -0,0 +1,1446 @@
...
+      "Supported charged_frag_types in the loaded weights  ['b_z1', 'b_z2', 'y_z1', 'y_z2', 'b_modloss_z1', 'b_modloss_z2', 'y_modloss_z1', 'y_modloss_z2']\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Excluding the modloss fragment types\n",
+    "model = pDeepModel(charged_frag_types=['b_z1', 'b_z2', 'y_z1', 'y_z2'])\n",
+    "model.load(new_path)\n",
+    "print(f\"Model Interface has charged_frag_types {model.charged_frag_types}\")\n",
+    "print(f\"Supported charged_frag_types in the loaded weights  {model.model.supported_charged_frag_types}\")"
```
Comment (GeorgWa) [inline nbs_trials/adapt_charged_fragtypes.ipynb:715]: The `model.model` interface feels a bit strange. could we improe the naming?

File: nbs_trials/adapt_charged_fragtypes.ipynb [hunk 4]
```diff
@@ -0,0 +1,1446 @@
...
+     "output_type": "stream",
+     "text": [
+      "Model Interface has requested charged_frag_types ['b_z1', 'b_z2', 'y_z1', 'y_z2', 'b_modloss_z1', 'b_modloss_z2', 'y_modloss_z1', 'y_modloss_z2']\n",
+      "Supported charged_frag_types in the loaded weights  ['b_z1', 'b_z2', 'y_z1', 'y_z2', 'b_modloss_z1', 'b_modloss_z2', 'y_modloss_z1', 'y_modloss_z2']\n"
+     ]
+    }
+   ],
+   "source": [
+    "model =  pDeepModel(\n",
+    "    charged_frag_types=['c_z1', 'c_z2', 'y_z1', 'y_z2', 'x_z1', 'x_z2', 'y_modloss_z1', 'y_modloss_z2'], # Will be overridden by the model weights\n",
```
Comment (GeorgWa) [inline nbs_trials/adapt_charged_fragtypes.ipynb:1012]: can we leave the `charged_frag_types` empty to emphasize this? 
`charged_frag_types = []`

File: nbs_trials/adapt_charged_fragtypes.ipynb [hunk 5]
```diff
@@ -0,0 +1,1446 @@
...
+       "3           0.0           0.0  \n",
+       "4           0.0           0.0  "
+      ]
+     },
+     "execution_count": 51,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
```
Comment (GeorgWa) [inline nbs_trials/adapt_charged_fragtypes.ipynb:1138]: Can we add an example that the order is automatically controlled by alphabase?

Even if the user requests a random order both the saved and predicted frag types are always in a defined order by `sort_charged_frag_types`.

File: nbs_trials/adapt_charged_fragtypes.ipynb [hunk 6]
```diff
@@ -0,0 +1,1446 @@
...
+   "metadata": {},
+   "source": [
+    "## User importing a new model "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Trining on new fragment types that were not part of the training of the original weights (New feature). \n",
```
Comment (GeorgWa) [inline nbs_trials/adapt_charged_fragtypes.ipynb:1271]: (nit) Trining

File: nbs_trials/adapt_charged_fragtypes.ipynb [hunk 7]
```diff
@@ -0,0 +1,1446 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 55,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Trying to predict when the requested fragment types are not supported by the pretrained model\n",
```
Comment (GeorgWa) [inline nbs_trials/adapt_charged_fragtypes.ipynb:1285]: can you actually print the `safe to predict` and `safe to train` property?
I feel like this would make it more understandible to the user

File: nbs_trials/adapt_charged_fragtypes.ipynb [hunk 8]
```diff
@@ -0,0 +1,1446 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.9.19"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```
Comment (GeorgWa) [inline nbs_trials/adapt_charged_fragtypes.ipynb:1446]: Really good notebook overall!!

I would suggest to include the test data so users can try it out themeselves. If you send me the or Magnus the data we can upload it.
Also I would suggest to label the sections like

```
1. MS2 Prediction 
1.1 Prediction with old weights format
1.2 Prediction with new weights format
2. MS2 Transfer Learning
...

```

File: peptdeep/model/ms2.py [hunk 9]
```diff
@@ -379,42 +379,64 @@ def forward(self, pred, target):
...
-        modloss_type="modloss",
         model_class: torch.nn.Module = ModelMS2Bert,
         device: str = "gpu",
+        mask_modloss: Optional[bool] = None,
+        override_from_weights: bool = False,
         **kwargs,  # model params
     ):
         super().__init__(device=device)
+        if mask_modloss is not None:
+            warnings.warn("mask_modloss is deprecated and will be removed in the future. To mask the modloss fragments, the charged_frag_types should not include the modloss fragments.")
```
Comment (GeorgWa) [inline peptdeep/model/ms2.py:414]: Does `mask_modloss` still work in this version?

File: peptdeep/model/ms2.py [hunk 10]
```diff
@@ -379,42 +379,64 @@ def forward(self, pred, target):
...
+    charged_frag_types : List[str]
+        Charged fragment types to predict
+    dropout : float, optional
+        Dropout rate, by default 0.1
+    model_class : torch.nn.Module, optional
+        Ms2 Model class, by default ModelMS2Bert
+    device : str, optional
+        Device to run the model, by default "gpu"
+    override_from_weights : bool, optional default False
+        Over ride the requested charged frag types when loading model from weights, this will
```
Comment (GeorgWa) [inline peptdeep/model/ms2.py:394]: Maybe make it more clear what this is doing.

Always update the requested charged fragment types from the model weights on loading. This allows to predict all fragment types supported by the weights even if the user doesn't know what fragments types are supported by the weights. Thereby, the model will always be in a safe to predict state.

File: peptdeep/model/ms2.py [hunk 11]
```diff
@@ -504,6 +530,41 @@ def _set_batch_predict_data(
                 predicts.reshape((-1, len(self.charged_frag_types))),
                 batch_df[["frag_start_idx", "frag_stop_idx"]].values,
             )
+    def _align_model_charged_frag_types(self):
+        """
+        Align the underlying model charged_frag_types with the interface charged_frag_types,
```
Comment (GeorgWa) [inline peptdeep/model/ms2.py:535]: are interface charged frag types the requested ones?

Comment (GeorgWa) [reply peptdeep/model/ms2.py:535]: I feel like the impact can be more clear. `align` sounds very wague.
Maybe something like `reshape_prediction_head` ?

File: peptdeep/model/ms2.py [hunk 12]
```diff
@@ -587,6 +655,10 @@ def predict(
         reference_frag_df=None,
         **kwargs,
     ) -> pd.DataFrame:
+        if not self._safe_to_predict:
```
Comment (GeorgWa) [inline peptdeep/model/ms2.py:658]: Please correct me, but shouldn't the most obvious solution be that users should select different (supported) charged frag types? I would mention this first and then using different weights and retraining.

File: peptdeep/model/ms2.py [hunk 13]
```diff
@@ -632,7 +704,49 @@ def bootstrap_nce_search(
...
+        if len(missing_keys) > 0:
+            logging.warn(
+                f"nn parameters {missing_keys} are MISSING while loading models in {self.__class__}"
+            )
+        if len(unexpect_keys) > 0:
+            logging.warn(
+                f"nn parameters {unexpect_keys} are UNEXPECTED while loading models in {self.__class__}"
+            )
+
+    def _update_model_state(self):
```
Comment (GeorgWa) [inline peptdeep/model/ms2.py:736]: could this lead to deviation of `_safe_to_predict` and the reality?
Do you think it would make sense to do this as part of a @property?

**Review by GeorgWa — COMMENTED** [review]
Really good and clean solution for the problem.
I would suggest to refine the notebook and PR description a bit to make it more clear 1) what is the problem 2) what does this PR implement 3) why is this a solution to the problem. This could be sections or headings etc.

File: peptdeep/model/ms2.py [hunk 14]
```diff
@@ -884,3 +995,31 @@ def calc_ms2_similarity(
 
     torch.cuda.empty_cache()
     return psm_df, metrics_describ
+
+
+def charged_frags_to_tensor(charged_frags: List[str]) -> torch.Tensor:
```
Comment (jalew188) [inline peptdeep/model/ms2.py:1000]: Yes, this looks very nice.

File: peptdeep/model/ms2.py (hunk 13 above)
Comment (GeorgWa) [reply peptdeep/model/ms2.py:733]: Ah I see 👍🏻

File: nbs_trials/test_predict_ms2.ipynb
Comment (GeorgWa) [inline nbs_trials/test_predict_ms2.ipynb:1]: Can you check if the table is rendered correctly?

File: nbs_trials/test_new_ms2_weight_format.ipynb
Comment (GeorgWa) [inline nbs_trials/test_new_ms2_weight_format.ipynb:1]: Can you add a reference where in alphabase the user finds the supported charge_frag_types?

File: nbs_trials/test_predict_ms2.ipynb
Comment (GeorgWa) [inline nbs_trials/test_predict_ms2.ipynb:1]: Can you check if the table is rendered correctly?

Comment (GeorgWa) [inline nbs_trials/test_predict_ms2.ipynb:1]: Can you check for `pretarined` ? I think you had a typo in some files :)

File: nbs_trials/test_transfer_ms2.ipynb
Comment (GeorgWa) [inline nbs_trials/test_transfer_ms2.ipynb:1]: Can you ask Magnus for help with the DataDownloader and uploading the `trainin_data_path = "../data/2oh_evidence_txt_0_batch_0.hdf"` ? Can be in a separate PR :)

**Review by GeorgWa — APPROVED** [review]
LGTM with some minor notebook changes
